python/sdk/merlin/observability/inference.py

```diff
@@ -1,111 +1,190 @@
+
+    def actual_score(self, actual_label: str, prediction_label: str) -> float:
+        """
+        Derive actual score from actual label and prediction label.
+        :param actual_label: Actual label.
+        :param prediction_label: Prediction label.
+        :return: actual score. Either 0.0 for the wrong prediction or 1.0 for the correct one.
+        """
+        if actual_label not in self.classes:
+            raise ValueError(f"Actual label must be one of the classes, got {actual_label}")
+        return 0.0 if actual_label != prediction_label else 1.0
 
 
 @dataclass_json
 @dataclass
 class RankingOutput:
+    """
+    Ranking model prediction output schema.
+
+    Attributes:
+        rank_column: Name of the column containing the rank of the prediction. Rank value is greater or equals to 1.
+        prediction_group_id_column: Name of the column containing the prediction group id.
+        relevance_score_column: Name of the column containing the relevance score, which serves as the ground truth
+            for the ranking model.
+    """
     rank_column: str
     prediction_group_id_column: str
-
-    @property
-    def column_types(self) -> Dict[str, ValueType]:
-        return {
-            self.rank_column: ValueType.INT64,
-            self.prediction_group_id_column: ValueType.STRING,
-        }
+    relevance_score_column: str
 
 
-@unique
 class InferenceType(Enum):
-    BINARY_CLASSIFICATION = 1
-    MULTICLASS_CLASSIFICATION = 2
-    REGRESSION = 3
-    RANKING = 4
+    """
+    Supported model types.
+    """
+    BINARY_CLASSIFICATION = auto()
+    SCORE_CATEGORICAL = auto()
+    REGRESSION = auto()
+    RANKING = auto()
 
 
 @dataclass_json
 @dataclass
 class InferenceSchema:
+    """
+    Inference schema for a Merlin model, which are required to extract the necessary model observability metrics
+    from either a prediction log or a Pandas dataframe, and send the metrics to an observability backend such as
+    Arize AI.
+
+    Attributes:
+        feature_types: Dictionary of feature names and their types.
+        type: Type of the model. One of binary_classification, multiclass_classification, regression, or ranking
+            attribute must be set depending on the model type.
+        binary_classification: Binary classification model prediction output schema.
+        score_categorical: Multiclass classification model prediction output schema.
+        regression: Regression model prediction output schema.
+        ranking: Ranking model prediction output schema.
+        prediction_id_column: Name of the column containing the prediction id, so that the ground truth can be joined
+            to the predictions.
+        tag_columns: List of column names containing tags for grouping the predictions.
+    """
+
     feature_types: Dict[str, ValueType]
     type: InferenceType
     binary_classification: Optional[BinaryClassificationOutput] = None
```

Comment:
instead of defining all fields for each model type can we make abstract class for each model_type_output and have preprocess as one of abstract method? it will help us later on when we add new model type

Reply:
One issue with using an abstract class for model_type_ouput, instead of a combination of inference type and fields for all model types, is that for the former, it is tricky to serialize to json then deserialize the json back to the concrete implementation of the abstract class. This will make it difficult to pass InferenceSchema / Model Type Output as a configuration to Flyte Workflows, or as a structure config (via OmegaConf library) in the observation publisher. We might have to implement a custom json serializer / deserializer for the class.

The other issue, is whether we want to expose preprocess method to the Merlin SDK users, consider that the method is intended to be used by the observation publishers (batch and online).